### DISCUSSION TITLE
Is there a method to retrieve all open windows of my app?

### DISCUSSION THREAD

**Discussion by fabianoriccardi** [microsoft/WindowsAppSDK#4502] [Q&A]
I was looking for a (possibly easy) method to retrieve all the opened `Window`s of my app. Currently, I can manually register all the opened windows in a global static List, but it would be great to have a native structure that keeps track of them.

I was looking for something similar to Application.Current.Windows in .NET4.8 (https://learn.microsoft.com/en-us/dotnet/desktop/wpf/app-development/how-to-get-all-windows-in-an-application?view=netframeworkdesktop-4.8).

**Comment by Yewzir** [comment]
@fabianoriccardi 

It's been a while, but I think the Windows API contains definitions for the methods ~`WindowEnum`~ and ~`ChildWindowEnum`~. There's certainly `GetWindow` too, but that method was less favorable to use for some reason, if I recall that somewhat correctly.

Edit: I looked it up and the methods are `EnumWindows` and `EnumChildWindows`, `GetWindow` was correct and is indeed less favorable to use than `EnumWindows`.

> 
> **[EnumWindows function (winuser.h)](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumwindows)**
> 
> Enumerates all top-level windows on the screen by passing the handle to each window, in turn, to an application-defined callback function. EnumWindows continues until the last top-level window is enumerated or the callback function returns FALSE.

> **[EnumChildWindows function (winuser.h)](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumchildwindows)**
> 
> Enumerates the child windows that belong to the specified parent window by passing the handle to each child window, in turn, to an application-defined callback function. EnumChildWindows continues until the last child window is enumerated or the callback function returns FALSE.

> 
> **[GetWindow function (winuser.h)](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindow)**
> 
> Retrieves a handle to a window that has the specified relationship ([Z-Order](https://learn.microsoft.com/en-us/windows/desktop/winmsg/window-features) or owner) to the specified window.

The header [`winuser.h`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/) contains several methods regarding windows, and you might certainly want to browse through them, because there may be more possibilities than the few I listed here.

Kind regards
@Yewzir

**Reply by fabianoriccardi** [reply]
Thanks for the reply @Yewzir, you are not too late.

I didn't exactly know these methods, but in the first post I had written "easy" because I was supposing that the low-level APIs would provide such feature, but I was looking for something "native" in Windows App SDK.

=> The developers may add it in the next releases.

**Reply by Yewzir** [reply]
@fabianoriccardi 

What programming language are you targeting specifically?

**Reply by fabianoriccardi** [reply]
C#
